server/api/student/student.controller.js

```diff
@@ -84,6 +87,123 @@ exports.outreachCounts = function(req, res) {
   });
 };
 
+function createNameAndSchoolPromise(results) {
+  return new Promise(function(resolve, reject) {
+    var promises = _.map(results, function(result) {
+      return new Promise(function(resolve, reject) {
+        function getTotals() {
+          var records = this.records;
+          var typeArr = _.map(records, 
+              function(rec) { return rec.type 
+            });
+          return _.countBy(typeArr, _.identity);
+        }
+        School.findById(result.school, 
+          function(err, school) {
+            if (err) reject(err);
+            result.school = school;
+          })
+          .then(function() {
+            Student.findById(result.student, 
+              function(err, student) {
+              if (err) reject(err);
+              result.student = student;
+            })
+          .then(function() {
+            result.totals = getTotals.call(result);
+            resolve(result);
+          });
+        });
+      });
+    });
+    Promise.all(promises).then(
+      function(val) { 
+        resolve(val);
+    });
+  });
+}
+
```

Comment:
I think something like this might work:
```
exports.outreachSummary = function(req, res) {
  var pipeline = [{
    $group: {
      _id: {student: '$student', type: '$type'},
      count: {$sum: 1}
    }
  }, {
    $group: {
      _id: '$_id.student',
      counts: {$push: {type: '$_id.type', count: '$count'}}
    }
  }];
  if (req.user.role === 'teacher') {
    pipeline.unshift({
      $match: {school: req.user.assignment}
    });
  }
  Outreach.aggregate(pipeline, function(err, results) {
    if (err) handleError(res, err);
    return res.status(200).json(results);
  });
};
```

You would need to populate the student before sending the results down to the client.